[LC-944] BlockVerifier 1.0 (Finalize)

metalg0su opened the pull request:
(Will be resolved after #511 is resolved)

# Goal
- Add BlockVerifier, which implements `DataVerifier` of LFT
- `BlockVerifier` inspects given `Block` and `Transacton`s.

zsaladin commented on loopchain/blockchain/blocks/v1_0/block_verifier.py:
```diff
@@ -1,9 +1,65 @@
+            tv = TransactionVerifier.new(tx.version, tx.type(), self.tx_versioner)
+            tv.verify(tx)
+
+    def _verify_something(self, block: 'Block'):
```
Can you give it more specific name to the method?

zsaladin commented on loopchain/blockchain/blocks/v1_0/block_verifier.py:
```diff
@@ -1,9 +1,65 @@
+            round_num=block.header.round,
+            invoke_request=invoke_request
+        )
+        # TODO: No more new_block comparison?
```
Timestamp, all root hashes

metalg0su replied:
Thank you. I understood.

I will apply your feedbacks right away.

zsaladin commented on loopchain/blockchain/invoke_result.py:
```diff
@@ -98,6 +98,23 @@ def _extract_validators(self, prev_block_votes: List[List[str]]) -> List[str]:
+            prev_block_hash=header.prev_hash,
+            timestamp=header.timestamp,
+            prev_votes=body.prev_votes,
+            tx_versioner=TransactionVersioner(),
```
TrasnactionVersioner must be composed in configuration.
It means that this has to use versionoer in Blockchain object.

metalg0su replied:
Thank you. I understood.

I will apply your feedbacks right away.

hyerim-kim committed:
[LC-944] Add BlockVerifier 1.0

- Add logic to verify the block version of `1.0` with invocation of the `iconservice`.
- Edit `reset_cache()` in `BlockVerifier`.
- Add some attributes in `InvokeData`.
- Edit the name of some variables.
- Add test.

Co-authored-by: Seongjun Ji <metalg0su@gmail.com>

hyerim-kim committed:
[LC-944] Apply feedback.

- Set `TransactionVersioner` from `Blockchain` object.

hyerim-kim committed:
[LC-944] Edit test cases.

hyerim-kim committed:
[LC-944] Edit some testcases.

hyerim-kim commented on loopchain/blockchain/blocks/v1_0/block_builder.py:
```diff
@@ -24,7 +25,6 @@ def __init__(self, tx_versioner: 'TransactionVersioner'):
         # Attributes that must be assigned
         self.validators: Optional[List[ExternalAddress]] = None
         self.next_validators: Optional[List[ExternalAddress]] = None
-        self.next_validators_hash: Optional[Hash32] = None
```
The `next_validators_hash` is generated after initializing related attributes. I referred above comments for categorizing attributes.

metalg0su replied:
👍

metalg0su commented on loopchain/blockchain/blocks/v1_0/block_builder.py:
```diff
@@ -56,18 +57,15 @@ def prev_receipts(self, prev_receipts):
-        # TODO In 'LFT', this method call is unnecessary. Delete after confirmation.
-        super().reset_cache()
-
+        self.block = None
+        self.hash = None
```
Can you explain why you removed `super().reset_cache()`, then add two lines?

hyerim-kim replied:
The purpose disappeared now to change codes, it should be recovered to the past.

metalg0su replied:
Sugguest to remove `reset_cache`.

metalg0su commented on loopchain/store/key_value_store_plyvel.py:
```diff
@@ -13,16 +13,11 @@
-from loopchain import configure_default as conf
-from loopchain.blockchain.transactions import TransactionSerializer, TransactionVersioner
-from loopchain.blockchain.types import Hash32
+from loopchain.blockchain.transactions import TransactionVersioner
```
It may be out of topic, but you can remove `KeyValueStorePlyvel.__tx_versioner`.
Then you can remove `TransactionVersioner` import line.

> not obligated. if you do not want to do this, I will handle it.
> Because it is caused by me. 😅

hyerim-kim replied:
👌

metalg0su commented on testcase/unittest/blockchain/blocks/v1_0/test_block_factory.py:
```diff
@@ -85,7 +85,7 @@ async def test_create_none_data(self, block_factory: v1_0.BlockFactory):
 
     @pytest.mark.asyncio
-    async def test_create_none_data(self, block_factory: v1_0.BlockFactory):
+    async def test_create_lazy_data(self, block_factory: v1_0.BlockFactory):
```
Thank you to fix typo.

metalg0su commented on loopchain/blockchain/votes/v1_0/vote.py:
```diff
@@ -157,10 +157,38 @@ def _deserialize(cls, **data) -> Vote:
+            sign_verifier.verify_hash(hash_, self._signature)
+        except Exception as e:
+            raise RuntimeError(f"Invalid vote signature. {self}\n"
+                               f"{e}")
```
Multi-line logs make logging much harder (imagine that you parse log lines by using `cat` or `grep` or something).

I suggest to use single line logging.

hyerim-kim replied:
I know what you mean, but single-line logging, which contains multi-line, is bad to read for me when I figure the problem case out. If you suggest to me how the log express better, I'll apply to code.

metalg0su replied:
I found LC-805 that contains:
- we decided to leave multiline logs in code lines.

I resolve this review.

metalg0su commented on loopchain/blockchain/blocks/v1_0/block_verifier.py:
```diff
@@ -1,25 +1,137 @@
+    def _verify_vote(self, vote: 'BlockVote'):
+        invoke_result = self._invoke_pool.get_invoke_data(vote.epoch_num, vote.round_num)
+
+        if invoke_result.state_hash != vote.state_hash:
+            raise RuntimeError(
+                f"Block({vote.height}, {vote.data_id.hex()}, "
+                f"StateRootHash({vote.state_hash.hex()}), "
+                f"Expected({invoke_result.state_hash.hex()}).")
+
+        if invoke_result.receipt_hash != vote.receipt_hash:
+            raise RuntimeError(
+                f"Block({vote.height}, {vote.data_id.hex()}, "
+                f"ReceiptRootHash({vote.receipts_hash.hex()}), "
+                f"Expected({invoke_result.receipts_hash.hex()}.")
+
+        if invoke_result.next_validators_hash != vote.next_validators_hash:
+            raise RuntimeError(
+                f"Block({vote.height}, {vote.data_id.hex()}, "
+                f"NextValidatorsHash({vote.next_validators_hash.hex()}), "
+                f"Expected({invoke_result.next_validators_hash.hex()}.")
```
LFT expects BlockVote to implement consensus_id.
And I created it by using 4 hashes which are consisted of hashes above.
And LFT passes vote verification only after consensus_id is verified (matched)

So I think this is redundant step.

metalg0su commented on loopchain/blockchain/blocks/v1_0/block_verifier.py:
```diff
@@ -1,25 +1,137 @@
+            raise e
+        return True
+
+    def _verify_vote(self, vote: 'BlockVote'):
```
LFT is responsible to do this, I think.

metalg0su replied:
It looks that LFT does not verify vote signature.
For now, it is needed.

> Please resolve the review if it is correct

metalg0su commented on loopchain/blockchain/blocks/v1_0/block_verifier.py:
```diff
@@ -1,25 +1,137 @@
+
+        if prev_votes:
+            for vote in prev_votes:
+                if vote and vote.version <= '0.5':
```
You intended: ?
- first `Block` of 1.0 contains Vote 0.5.

hyerim-kim replied:
Yes, for once I consider the case the first block of the `LFT` contains votes of a block by the `Siever`.

metalg0su commented on loopchain/blockchain/blocks/v1_0/block_verifier.py:
```diff
@@ -1,25 +1,137 @@
-from loopchain.blockchain.transactions import TransactionVersioner
+from loopchain import configure_default as conf
+from loopchain import utils
+from loopchain.blockchain import ExternalAddress
```
I think this line could raise import error, because module `blockchain` does not import module `blockchain.types` (Indirect import)

Possible solutions:
- `from loopchain.blockchain.types import ExternalAddress`
- import blockchain.types at `blockchain.__init__.py`
- Leave it as it is.

hyerim-kim replied:
Yes, I guess indirect imports should be refactoring all parts of this project.

metalg0su replied:
> Resolve the conversation if you want.

metalg0su commented on loopchain/blockchain/blocks/v1_0/block_factory.py:
```diff
@@ -60,7 +60,7 @@ async def create_data(self, data_number: int, prev_id: bytes, epoch_num: int, ro
 
         pre_invoke_response: 'PreInvokeResponse' = self._invoke_pool.prepare_invoke(
-            block_height=data_number,
+            block_height=data_number-1,
```
Can you explain why pre_invoke height is changed?

hyerim-kim replied:
It will be recovered.

metalg0su commented on loopchain/blockchain/blocks/v1_0/block_verifier.py:
```diff
@@ -1,25 +1,137 @@
+        except Exception as e:
+            raise RuntimeError(f"Block({block.header.height}, {block.header.hash.hex()}, Invalid Signature {e}")
+
+    def _process_next_validators(self, block: 'Block', builder: BlockBuilder, next_validators: dict):
```
This method is not used..?

metalg0su commented on loopchain/blockchain/invoke_result.py:
```diff
@@ -176,6 +176,8 @@ def __init__(self,
         self._height = height
         self._receipts: list = receipts
         self._state_hash: Optional[Hash32] = state_root_hash
+        self._validators_hash = validators_hash
+        self._next_validators = next_validators_origin
```
are they used?

metalg0su commented on loopchain/blockchain/invoke_result.py:
```diff
@@ -229,7 +239,9 @@ def new(cls, epoch_num, round_num,
-        next_validators_info: Optional[dict] = invoke_result.get("prep")
+        current_reps_hash = invoke_result.get('currentRepsHash')
+        current_validators_hash = current_reps_hash if current_reps_hash else current_validators_hash
+        next_validators_info: Optional[dict] = invoke_result.get("prep", {})
```
need to set empty `dict`?

metalg0su commented on loopchain/blockchain/votes/v1_0/vote_factory.py:
```diff
@@ -70,9 +70,9 @@ async def create_vote(self, data_id: bytes, commit_id: bytes, epoch_num: int, ro
+        timestamp = utils.get_time_stamp()
         signature = self._get_signature(
-            height=0,
+            height=-1,
```
> Possible related?: https://github.com/icon-project/loopchain/pull/512/files#r464295994

Can you explain why you set the height `-1`

hyerim-kim replied:
It will be recovered.

metalg0su commented on testcase/unittest/blockchain/blocks/v1_0/test_block_verifier.py:
```diff
@@ -0,0 +1,242 @@
+            # THEN verification failed
+            await block_verifier.verify(prev_block, block)
+
+    @pytest.mark.xfail(reason="Validate reps on BlockVerifier or in Epoch?")
+    async def test_reps(self, block_verifier, create_block):
+        assert 0
+
+    @pytest.mark.xfail(reason="What should be tested on verify_prev_votes?")
+    async def test__verify_prev_votes(self, block_verifier, create_block):
+        assert 0
```
Please complete those test case if we decided to verify about validators and prev_votes.

hyerim-kim committed:
[LC-944] Apply feedbacks.

metalg0su commented on loopchain/blockchain/invoke_result.py:
```diff
@@ -209,6 +209,10 @@ def height(self) -> int:
     def state_hash(self) -> Hash32:
         return self._state_hash
 
+    @property
+    def validators_hash(self) -> Hash32:
+        return self._validators_hash
+
```
1. 
`InvokeData` has no `_validators_hash`.
Please check.

2.
It looks that this prop has no outer usage.

metalg0su commented on loopchain/blockchain/invoke_result.py:
```diff
@@ -260,10 +266,10 @@ def prepare_invoke(self, block_height: int, block_hash: Hash32) -> PreInvokeResp
         return PreInvokeResponse.new(pre_invoke_result)
 
-    def invoke(self, block: 'Block') -> InvokeData:
+    def invoke(self, block: 'Block', tx_versioner: 'TransactionVersioner') -> InvokeData:
```
> This is not a review.

I sense that "invoke needs tx_versioner" is a kind of.. a sign for improvement.

But I understand what you want, and it is the best way to supply tx_versioner here I think.
I have no idea to invoke without the versioner. 😢

metalg0su commented on loopchain/blockchain/blocks/v1_0/block_verifier.py:
```diff
@@ -1,25 +1,106 @@
+    def _verify_prev_votes_by_siever(self, block: 'Block', prev_reps: Sequence['ExternalAddress'], any_vote):
+        header: 'BlockHeader' = block.header
+        body: 'BlockBody' = block.body
+        votes_class = v0_5.BlockVotes if any_vote.version else v0_1a.BlockVotes
```
class `Votes` provides useful method called `get_block_votes_class`. How about using it?
This helps avoiding import lines of Votes `v0_1a, v0_5`.

metalg0su commented on loopchain/blockchain/blocks/v1_0/block_verifier.py:
```diff
@@ -1,25 +1,106 @@
+from loopchain import utils
+from loopchain.blockchain.exception import BlockVersionNotMatch, TransactionOutOfTimeBound
+from loopchain.blockchain.transactions import TransactionVerifier, TransactionVersioner
+from loopchain.blockchain.types import ExternalAddress
```
You are using `ExternalAddress` as Type hint.
Options are:
- Locate under `if TYPE_CHECKING`
- Remove quotes at type hint line.

metalg0su commented:
I approved!

But I suggest that you merge this PR after all other latest PRs including `feature-10[46-50]` are merged.
Because those PRs may raise some problems if Block Verification is merged.
So merge basic functional PRs first, then verification.

yakkle commented:
stale
